File: spoj/3486-Elim.cpp
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include <sys/time.h>
#include <sys/resource.h>
#include<unistd.h>
//#include"memu.c"
using namespace std;




const int max = 1e6;
const int c_max= 1030;

char line[max];
int vi_l[c_max],v_ind[c_max];

void retirar(int* vl, int* vi,int l,int m){
	
	if ( l >=2){
		int* v1=(int*)malloc(l*sizeof(int));
		for (int i=0;i<l;i++){
			v1[i]= vl[vi[m+i]-1];
		}
		m+=l;
		retirar(v1,vi,l/2,m);
		free(v1);
	}else{
		printf("%d\n",vl[vi[m]-1]);
	}
}

//tokenizer
void tknzr(char* vl,int* vi,int* s){
	char* token;	
/*	char* a= (char*) malloc(max*sizeof(char));
	strcpy(a,vl);
*/
	while ((token = strsep(&vl," "))!= NULL && *s<1025){
			vi[(*s)++]=atoi(token);
	}
//	free(a);
}

int  main(){
	while(scanf("%[^\n]\n",line) && strcmp(line,"0")!=0){
		int size=0;
		tknzr(line,vi_l,&size);
		scanf("%[^\n]\n",line);
		int aux=0;
		tknzr(line,v_ind,&aux);
		if (size==1){
				printf("%d\n",vi_l[0]);
		}else
			printf(" ");
			retirar(vi_l,v_ind,size/2,0);
	}
//	free(line);
/*	size_t currentSize = getCurrentRSS( );                                                                                                      
                printf("usage: %lu \n",currentSize);
	printf("usage: %d \n",getmem());       		
*/
}

File: spoj/3486-Elim.py
from math import sqrt
def retirar(vl,vi,l):
	if l!=2:
		v1,v2=[],[]
		for i in xrange(l):
			v1.append(vl[vi[i]-1])
		v2=vi[l:]	
		retirar(v1,v2,l/2)
	else:	
		v1,v2=[],[]
		v1.append(vl[vi[0]-1])
		v1.append(vl[vi[1]-1])
		print v1[vi[2]-1]	


line= raw_input()
while line!='0':
	v_l=line.split()
	vi_l,vi_ind=[],[]
	for i in xrange(len(v_l)):
		vi_l.append(int(v_l[i]))
	v_ind= raw_input().split()
	for i in xrange(len(v_ind)):
		vi_ind.append(int(v_ind[i]))
	#print vi_l
	#print vi_ind
	if len(vi_l)/2 !=1:
		retirar(vi_l,vi_ind,len(vi_l)/2)
	else:
		print vi_l[vi_ind[0]-1]
	line=raw_input()
	
